MaxText/convert_llama_ckpt.py

```diff
@@ -0,0 +1,346 @@
+  layer_weight['pre_self_attention_layer_norm']['scale'] = np.transpose(
+                                    layer_weight['pre_self_attention_layer_norm']['scale'],
+                                    axes=(1, 0))
+  layer_weight['post_self_attention_layer_norm']['scale'] = np.transpose(
+                                    layer_weight['post_self_attention_layer_norm']['scale'],
+                                    axes=(1, 0))
+
+  jax_weights['decoder']['decoder']['mlp'] = layer_weight['mlp']
+  jax_weights['decoder']['decoder']['pre_self_attention_layer_norm'] = layer_weight['pre_self_attention_layer_norm']
+  jax_weights['decoder']['decoder']['post_self_attention_layer_norm'] = layer_weight['post_self_attention_layer_norm']
+
+  #convert all weights to jax.numpy
+  jax_weights = jax.tree_map(jnp.array, jax_weights)
+
+  print(f"jax_weights = {jax_weights}")
+
+  #dummy configs for the checkpoint_manager
+  step_number_to_save_new_ckpt = 0
+
+  base_output_directory="base_output_directory=dummy_base_output_dir"
+  base_num_decoder_layers=f"base_num_decoder_layers={base_num_decoder_layers}"
+  base_num_query_heads = f"base_num_query_heads={base_num_query_heads}"
+  base_num_kv_heads = f"base_num_kv_heads={base_num_kv_heads}"
+  head_dim = f"head_dim={head_dim}"
+  async_checkpointing="async_checkpointing=False"
+  enable_dropout="enable_dropout=False"
+
+  vocab_size=f"vocab_size={vocab_size}"
+  base_emb_dim=f"base_emb_dim={base_emb_dim}"
+  base_mlp_dim=f"base_mlp_dim={base_mlp_dim}"
+  model_name = "model_name=llama2-7b"
+
+  commandline_args = ["",
+                      "MaxText/configs/base.yml",
+                      "steps=20", base_emb_dim, base_mlp_dim, base_num_query_heads, base_num_kv_heads,
+                      head_dim, vocab_size, base_num_decoder_layers,
+                      model_name, async_checkpointing,
+                      base_output_directory, enable_dropout]
+  print(f"commandline_args = {commandline_args}")
+  pyconfig.initialize(commandline_args)
+  config = pyconfig.config
+  init_rng, _ = random.split(random.PRNGKey(config.init_weights_seed), 2)
+  devices_array = max_utils.create_device_mesh(config)
+  mesh = Mesh(devices_array, config.mesh_axes)
+  model = Transformer(config, mesh)
+
+
+  checkpoint_manager = checkpointing.create_orbax_checkpoint_manager(
+      maxtext_model_path,
+      config.enable_checkpointing,
+      config.async_checkpointing,
+      config.checkpoint_period
+  )
+
+
+  if save_training_ckpt:
+    print("Creating a training state")
+    learning_rate_schedule = max_utils.create_learning_rate_schedule(config)
+    tx = maxtext_utils.get_optimizer(config, learning_rate_schedule)
+    state_new, _ = max_utils.setup_training_state(
+      model, tx, config, init_rng, mesh, checkpoint_manager
+      )
+  else:
+    print("Creating a decoding state")
+    state_new, _ = max_utils.setup_decode_state(
```

Comment:
Can't we always save the decoding state? Like we're just hallucinating the trainstate?

Reply:
(The decoding state should work during training as a parameter state, right?)

Reply:
modified according to offline discussion - thank you!